### ISSUE TITLE
Tilt becomes sluggish after log size grows very large

### ISSUE THREAD

**Issue by andymartin-sch** [tilt-dev/tilt#3359]
Our developers will occasionally leave Tilt running for a long time and/or launch services in DEBUG mode and a lot of content will be logged. After a while, so much content is logged that Tilt becomes sluggish and other issues like https://github.com/tilt-dev/tilt/issues/3194 appear.

It would be nice if [k8s_resource](https://docs.tilt.dev/api.html#api.k8s_resource) was extended to allow developers to set a maximum number of log lines for Tilt to retain on a per-service basis (and/or if there was a global flag to set a default to be applied across all services).

This would potentially also resolve some of the drivers behind https://github.com/tilt-dev/tilt/issues/1885

**Comment by nicks** [comment]
Hmmm...Tilt already puts a maximums on both the log store (how many logs are stored) and the rendering engine (how many logs are displayed). It also has a protocol for streaming logs incrementally. So I'm curious what part of this is falling down.

Do you have a sense of what part is sluggish? Is it the Tilt process itself? Is it the browser? If you close the browser window, does the issue go away? etc.

**Comment by andymartin-sch** [comment]
> Do you have a sense of what part is sluggish? Is it the Tilt process itself? Is it the browser? If you close the browser window, does the issue go away? etc.

I think it's Tilt itself as `tilt dump webview` can take quite a while to complete.

The browser UI experience is that it can take a long time (~30s) for the Tilt UI to load, although switching between services after the initial loading is done seems to still be pretty snappy.

I repro'd things with a service that just logged a bunch of stuff and this is how `tilt dump webview` slowed down as the logs grew:
```
$ time tilt dump webview | wc
   3584   13264  225508

real    0m0.077s
user    0m0.064s
sys     0m0.019s

$ time tilt dump webview | wc
 810325 1631657 23574473
 
real    0m3.892s
user    0m1.584s
sys     0m0.425s

$ time tilt dump webview | wc
1614693 3243812 46831984

real    0m11.490s
user    0m4.653s
sys     0m6.437s

$ time tilt dump webview | wc
5628232 11271986 163124286

real    1m26.399s
user    0m20.864s
sys     1m9.538s
```

In the "real world" case that the engineer sent in, his webview was roughly as large as the final example above:
```
$ tilt dump webview | wc
 3342358 7107448 110867384
```

**Comment by nicks** [comment]
thanks! that's helpful. ya, i think the original pageload downloads all the existing logs. we should try to tweak this a bit so that it either only downloads the first window, or limits the initial pageload, or something else

**Comment by nicks** [comment]
I ratcheted down the maximum log threshold a bit in https://github.com/tilt-dev/tilt/pull/3368, which I think will fix the immediate problem.

I also filed an enhancement request for the more long-term fix https://github.com/tilt-dev/tilt/issues/3369

I think @landism is planning to do a release today with this fix. I'm going to close this issue for now, but please re-open if it's still feeling too sluggish in the new version

**Comment by andymartin-sch** [comment]
Thanks! Will try it out and get back to you sometime next week

**Comment by andymartin-sch** [comment]
The engineer said there was a slight improvement. It's still not blocking anything critical, but it sounds like there's still something to be gained from the long-term fix in https://github.com/tilt-dev/tilt/issues/3369

Let me know if you want any more info from our end

**Comment by nicks** [comment]
Hmmm...if it's only a slight improvement, that's concerning, and makes me worry that we've misdiagnosed the issue (i would have expected a large improvement). cc @jazzdan for triage

**Comment by andymartin-sch** [comment]
Ok I experimented a bit more and I do see the webview size and load-time getting capped at a lower value than before:
```
$ time tilt dump webview | wc 
 300229  637963 7918372

real    0m2.176s
user    0m0.555s
sys     0m0.055s

$ time tilt dump webview | wc 
 375789  798528 9911261

real    0m3.759s
user    0m1.014s
sys     0m0.820s

$ time tilt dump webview | wc 
 295981  628936 7806358

real    0m2.173s
user    0m0.581s
sys     0m0.406s
```

When I load the Tilt UI, it can still take ~3-5s to do the initial load, which still "feels" slow and is likely what the engineer was referring to.

Having it capped like this is still a big improvement, so up to you all whether it's worth dropping the limit again to something lower or just waiting for https://github.com/tilt-dev/tilt/issues/3369.

**Comment by jazzdan** [comment]
@andymartin-sch I wonder if this is made worse by the fact that in your setup the Tilt web UI is reaching out over a network. For a long running Tilt instance with a lot of logs each websocket message can be several megabytes.

**Comment by andymartin-sch** [comment]
> @andymartin-sch I wonder if this is made worse by the fact that in your setup the Tilt web UI is reaching out over a network. For a long running Tilt instance with a lot of logs each websocket message can be several megabytes.

@jazzdan yeah that could be a factor. I can try again once https://github.com/tilt-dev/tilt/pull/3395 gets released to see if makes a difference. thanks!
